lib/auth/oidc.go

```diff
@@ -36,38 +37,103 @@ import (
 	"github.com/gravitational/trace"
 )
 
+func (s *AuthServer) getOrCreateOIDCClient(conn services.OIDCConnector) (*oidc.Client, error) {
+	client, err := s.getOIDCClient(conn)
+	if err == nil {
+		return client, nil
+	}
+	if !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+	return s.createOIDCClient(conn)
+}
+
 func (s *AuthServer) getOIDCClient(conn services.OIDCConnector) (*oidc.Client, error) {
 	s.lock.Lock()
 	defer s.lock.Unlock()
 
-	config := oidc.ClientConfig{
-		RedirectURL: conn.GetRedirectURL(),
-		Credentials: oidc.ClientCredentials{
-			ID:     conn.GetClientID(),
-			Secret: conn.GetClientSecret(),
-		},
-		// open id notifies provider that we are using OIDC scopes
-		Scope: utils.Deduplicate(append([]string{"openid", "email"}, conn.GetScope()...)),
+	clientPack, ok := s.oidcClients[conn.GetName()]
+	if !ok {
+		return nil, trace.NotFound("connector %v is not found", conn.GetName())
 	}
 
-	clientPack, ok := s.oidcClients[conn.GetName()]
+	config := oidcConfig(conn)
 	if ok && oidcConfigsEqual(clientPack.config, config) {
 		return clientPack.client, nil
 	}
+
 	delete(s.oidcClients, conn.GetName())
+	return nil, trace.NotFound("connector %v has updated the configuration and is invalidated", conn.GetName())
 
+}
+
+func (s *AuthServer) createOIDCClient(conn services.OIDCConnector) (*oidc.Client, error) {
+	config := oidcConfig(conn)
 	client, err := oidc.NewClient(config)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	client.SyncProviderConfig(conn.GetIssuerURL())
+	ctx, cancel := context.WithTimeout(context.Background(), defaults.WebHeadersTimeout)
+	defer cancel()
+
+	go func() {
+		defer cancel()
+		client.SyncProviderConfig(conn.GetIssuerURL())
+	}()
+
+	select {
+	case <-ctx.Done():
+	case <-s.closeCtx.Done():
+		return nil, trace.ConnectionProblem(nil, "auth server is shutting down")
+	}
+
+	// Canceled is expected in case if sync provider config finishes faster
+	// than the deadline
+	if ctx.Err() != nil && ctx.Err() != context.Canceled {
+		var err error
+		if ctx.Err() == context.DeadlineExceeded {
+			err = trace.ConnectionProblem(err,
+				"failed to reach out to oidc connector %v, most likely URL %q is not valid or not accessible, check configuration and try to re-create the connector",
+				conn.GetName(), conn.GetIssuerURL())
+		} else {
+			err = trace.ConnectionProblem(err,
+				"unknown problem with connector %v, most likely URL %q is not valid or not accessible, check configuration and try to re-create the connector",
+				conn.GetName(), conn.GetIssuerURL())
+		}
+		s.EmitAuditEvent(events.UserSSOLoginFailure, events.EventFields{
```

Comment:
Can you log the error here as well? It will help in the situation where we just get a dump of process logs from a customer.

Reply:
So I"m trying to differentiate server errors vs user/config errors. I hope we can train admins to get all the info they need from audit logs, as this makes much more sense.